### PR TITLE
Transient to Scoped registration for core services

### PR DESCRIPTION
**Pull request by dv00d00**
I am not sure about this one, seems like nothing prevents scoped registration unless I've missed something.

This should help reduce allocations in the scenario when detection services are injected by user code in several places during request lifetime

### CODE REVIEW

**Comment by dv00d00** [comment]
Benchmark code
```csharp
    [ShortRunJob]
    [MemoryDiagnoser]
    public class DIBench
    {
        [Params(1,5,10)]
        public int N;
        
        private IServiceProvider _provider;

        [GlobalSetup(Target = "GoScoped")]
        public void GlobalSetupScoped()
        {
            var sc = new ServiceCollection();

            sc.AddScoped<IHttpContextAccessor>(_ => new HttpContextAccessor
            {
                HttpContext = new DefaultHttpContext
                {
                    Request =
                    {
                        Headers =
                        {
                            ["User-Agent"] = "Mozilla/5.0 (iPad; CPU OS 13_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/81.0.4044.124 Mobile/15E148 Safari/604.1",
                        }
                    }
                }
            });

            sc.AddSingleton(new DetectionOptions());
            sc.TryAddScoped<IUserAgentService, UserAgentService>();
            sc.TryAddScoped<IDeviceService, DeviceService>();
            sc.TryAddScoped<IEngineService, EngineService>();
            sc.TryAddScoped<IPlatformService, PlatformService>();
            sc.TryAddScoped<IBrowserService, BrowserService>();
            sc.TryAddScoped<ICrawlerService, CrawlerService>();
            sc.TryAddScoped<IDetectionService, DetectionService>();
            
            _provider = new DefaultServiceProviderFactory().CreateServiceProvider(sc);
        }
        
        [GlobalSetup(Target = "GoTransient")]
        public void GlobalSetupTransient()
        {
            var sc = new ServiceCollection();

            sc.AddScoped<IHttpContextAccessor>(_ => new HttpContextAccessor
            {
                HttpContext = new DefaultHttpContext
                {
                    Request =
                    {
                        Headers =
                        {
                            ["User-Agent"] = "Mozilla/5.0 (iPad; CPU OS 13_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/81.0.4044.124 Mobile/15E148 Safari/604.1",
                        }
                    }
                }
            });

            sc.AddSingleton(new DetectionOptions());
            sc.TryAddTransient<IUserAgentService, UserAgentService>();
            sc.TryAddTransient<IDeviceService, DeviceService>();
            sc.TryAddTransient<IEngineService, EngineService>();
            sc.TryAddTransient<IPlatformService, PlatformService>();
            sc.TryAddTransient<IBrowserService, BrowserService>();
            sc.TryAddTransient<ICrawlerService, CrawlerService>();
            sc.TryAddTransient<IDetectionService, DetectionService>();
            
            _provider = new DefaultServiceProviderFactory().CreateServiceProvider(sc);
        }

        [Benchmark]
        public void GoScoped()
        {
            using var serviceScope     = _provider.CreateScope();
            for (int i = 0; i < N; i++)
            {
                var detectionService = serviceScope.ServiceProvider.GetRequiredService<IDetectionService>();

            }
        }

        [Benchmark(Baseline = true)]
        public void GoTransient()
        {
            using var serviceScope     = _provider.CreateScope();
            for (int i = 0; i < N; i++)
            {
                var detectionService = serviceScope.ServiceProvider.GetRequiredService<IDetectionService>();
                
            }
        }
    }
```

Results

|      Method |  N |      Mean |      Error |    StdDev | Ratio |    Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------ |--- |----------:|-----------:|----------:|------:|---------:|------:|------:|----------:|
|    GoScoped |  1 |  70.61 us |   2.216 us |  0.121 us |  0.87 |  15.7471 |     - |     - |  48.36 KB |
| GoTransient |  1 |  80.86 us |   0.969 us |  0.053 us |  1.00 |  20.2637 |     - |     - |  62.16 KB |
|             |    |           |            |           |       |          |       |       |           |
|    GoScoped |  5 |  69.40 us |   3.221 us |  0.177 us |  0.17 |  15.7471 |     - |     - |  48.36 KB |
| GoTransient |  5 | 419.55 us | 500.165 us | 27.416 us |  1.00 |  99.6094 |     - |     - | 305.33 KB |
|             |    |           |            |           |       |          |       |       |           |
|    GoScoped | 10 |  71.91 us |  18.347 us |  1.006 us |  0.09 |  15.7471 |     - |     - |  48.36 KB |
| GoTransient | 10 | 789.81 us |  61.802 us |  3.388 us |  1.00 | 198.2422 |     - |     - | 609.29 KB |

**Comment by wangkanai** [comment]
Transient is prepared over scoped, because the user can change their preference using a session scoped. [Dependency injection service lifetimes ](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-3.1#register-additional-services-with-extension-methods). 

Please correct me if my understanding it wrong here. Maybe you can look deeper in the matter in the follow code that make me think using transient is a better idea over scoped.

**Comment by dv00d00** [comment]
I am taking a closer look for sure, will respond with (hopefully) detailed investigation

**Comment by dv00d00** [comment]
I took a closer look at `ResponsiveService`.
It would not capture `HttpContext.Session` mutation after initial creation for the current request, but will capture it in a subsequent one. This is subpar experience.

Some options to fix it but minimise memory footprint:

- Leave `ResponsiveService` transient, all the heavy lifting is done in other services so this would just allocate class on the heap (96 bytes or so)
- Make `ResponsiveService` scoped and use lazy evaluation for the `View` property, so instead of calculating it upfront, calculate on demand with the current values from `HttpContext.Session`

I personally prefer the last one.

Let me know what you think and I will update PR.

**Comment by wangkanai** [comment]
`ResponsiveService` being scoped or transient would need to further investigation for sure.

But I think other services does not need to be transient, because the client device should change user agent (except for testing in the browser). What do you think about this ideas.

**Comment by dv00d00** [comment]
Updated `ResponsiveService` to honor session mutations and keep `Scoped` registration.

File: src/Services/ResponsiveService.cs [hunk 1]
```diff
@@ -12,24 +12,24 @@ namespace Wangkanai.Detection.Services
 {
     public class ResponsiveService : IResponsiveService
     {
-        public Device View { get; }
+        public Device View => PreferView(_context, _defaultView);
```
Comment (wangkanai) [inline src/Services/ResponsiveService.cs:15]: Interesting

File: src/Services/ResponsiveService.cs [hunk 2]
```diff
@@ -12,24 +12,24 @@ namespace Wangkanai.Detection.Services
...
-            if (options is null)
-                options = new DetectionOptions();
 
-            _context = accessor.HttpContext;
+            options = options ?? new DetectionOptions();
 
-            View = DefaultView(deviceService.Type, options.Responsive);
-            View = PreferView(_context, View);
+            _context     = accessor.HttpContext;
+            _defaultView = DefaultView(deviceService.Type, options.Responsive);
```
Comment (wangkanai) [inline src/Services/ResponsiveService.cs:32]: Very interesting, created a default view base from configuration. Which is application wide singleton lifecycle. where you made it read only.

**Comment by wangkanai** [comment]
Thank you very much, this changes look very promising and would create another great performance increase.

**Pull request merged**